File: day-1/go/main.go
package main

import (
    "fmt"
    "os"
    "strings"
    "strconv"
    "sort"
)

func main() {
    content_bytes, err := os.ReadFile("data.txt")
    if err != nil {
        fmt.Println(err)
    }

    blocks := strings.Split(string(content_bytes), "\n\n")

    // Now to calculate everything by nesting it in ugly for-loops because Go sucks
    max_elf_cals := 0
    for _, elf_block := range blocks {
        elf_cals := 0
        for _, elf_block := range strings.Split(elf_block, "\n") {
            if elf_block != "" {
                elf_cal_block, err := strconv.Atoi(elf_block)
                if err != nil {
                    fmt.Println(err)
                }
                elf_cals = elf_cals + elf_cal_block
                if elf_cals > max_elf_cals {
                    max_elf_cals = elf_cals
                }
            }
        }
    }

    fmt.Println("The Elf with the most cals is carrying", max_elf_cals, "calories.")

    elf_cal_counter := []int{}
    for _, elf_block := range blocks {
        elf_cals := 0
        for _, elf_block := range strings.Split(elf_block, "\n") {
            if elf_block != "" {
                elf_cal_block, err := strconv.Atoi(elf_block)
                if err != nil {
                    fmt.Println(err)
                }
                elf_cals = elf_cals + elf_cal_block
            }
        }
        elf_cal_counter = append(elf_cal_counter, elf_cals)
    }
    sort.Ints(elf_cal_counter)

    // Next we calculate the sum using *another* for-loop because Go sucks
    top_three_cal_total := 0
    for _, cals := range elf_cal_counter[len(elf_cal_counter)-3:] {
        top_three_cal_total = top_three_cal_total + cals
    }

    fmt.Println("The top three calorie carrying elves have a total", top_three_cal_total, "calories.")

}

File: day-1/rust/src/main.rs
use std::include_str;

fn main() {
    static DATA: &str = include_str!("data.txt");

    let max_cal_count = cal_counter(&DATA);
    println!("The top Elf is carrying the most calories at {}", max_cal_count);

    let sorted_max_cal_elf = combined_top_three(DATA);
    println!("The top three biggest calorie carrying elves have a combined {} cals", sorted_max_cal_elf);
}

fn cal_counter(data: &str) -> i32 {
    data
    .split("\n\n")
        .map(|r|
             r
             .lines()
             .map(|e|
                  e
                  .parse()
                  .unwrap_or(0))
             .sum::<i32>())
        .max()
        .unwrap()
}

fn combined_top_three(data: &str) -> i32 {
    let out = &mut data
        .split("\n\n")
        .map(|r|
             r
             .lines()
             .map(|e|
                  e
                  .parse()
                  .unwrap_or(0))
             .sum::<i32>())
        .collect::<Vec<i32>>();
    
    out.sort_by(|x, y| y.cmp(&x));

    out[..3].iter().sum()
}

#[test]
fn example_tests() {
    static TEST_DATA: &str =
r#"1000
2000
3000

4000

5000
6000

7000
8000
9000

10000"#;

    assert_eq!(cal_counter(&TEST_DATA), 24000);
    assert_eq!(combined_top_three(TEST_DATA), 45000);
}
